### PR TITLE
test mac release after NIC deletion

### PR DESCRIPTION
**Pull request by alonSadan**
test that when a VM NIC is remmoved , it's
mac address is released and can be used as
a custom mac for another VM configuration.

### CODE REVIEW

File: tests/virtual_machines_test.go [hunk 1]
```diff
@@ -450,6 +449,58 @@ var _ = Describe("Virtual Machines", func() {
 				Expect(err).ToNot(HaveOccurred())
 			})
 		})
+
+		Context("When a VM's NIC is removed and a new VM is created with the same MAC", func() {
+			It("should successfully release the MAC and the new Vm should be created with no errors ", func() {
```
Comment (SchSeba) [inline tests/virtual_machines_test.go]: remove the space at the end please

File: tests/virtual_machines_test.go [hunk 2]
```diff
@@ -450,6 +449,63 @@ var _ = Describe("Virtual Machines", func() {
...
+					}
+					return nil
+
+				}, 50*time.Second, 5*time.Second).Should(Not(HaveOccurred()), "failed to update client")
+
+				By("after the VM's NIC has been deleted")
+				Expect(updatedVM.Spec.Template.Spec.Domain.Devices.Interfaces[0].MacAddress).To(Equal(mac2.String()),
+					"kubemacppool should assign the same mac as before the update")
+				Eventually(func() error {
+					return testClient.VirtClient.Create(context.TODO(), newVm)
```
Comment (SchSeba) [inline tests/virtual_machines_test.go]: here you need to create a new vm not the one you already use.

Comment (alonSadan) [reply tests/virtual_machines_test.go]: but I think the creation of "newVm" should fail, so it is not really used, isn't it?

File: tests/virtual_machines_test.go [hunk 3]
```diff
@@ -450,6 +449,63 @@ var _ = Describe("Virtual Machines", func() {
...
+					if len(updatedVM.Spec.Template.Spec.Domain.Devices.Interfaces) != 1 {
+						return fmt.Errorf("failed to delete the VM's interface")
+					}
+					return nil
+
+				}, 50*time.Second, 5*time.Second).Should(Not(HaveOccurred()), "failed to update client")
+
+				By("after the VM's NIC has been deleted")
+				Expect(updatedVM.Spec.Template.Spec.Domain.Devices.Interfaces[0].MacAddress).To(Equal(mac2.String()),
+					"kubemacppool should assign the same mac as before the update")
```
Comment (yossisegev) [inline tests/virtual_machines_test.go]: typo: kubemacppool (you have 1 extra 'p').

Comment (alonSadan) [reply tests/virtual_machines_test.go]: Thanks:)

**Comment by SchSeba** [comment]
please rebase this PR

File: tests/virtual_machines_test.go [hunk 4]
```diff
@@ -489,6 +489,63 @@ var _ = Describe("Virtual Machines", func() {
 				Expect(err).ToNot(HaveOccurred())
 			})
 		})
+
+		Context("When a VM's NIC is removed and a new VM is created with the same MAC", func() {
+			It("should successfully release the MAC and the new Vm should be created with no errors", func() {
```
Comment (SchSeba) [inline tests/virtual_machines_test.go]: Vm -> VM

File: tests/virtual_machines_test.go [hunk 5]
```diff
@@ -489,6 +489,63 @@ var _ = Describe("Virtual Machines", func() {
...
+				Expect(err).ToNot(HaveOccurred())
+				mac2, err := net.ParseMAC(vm.Spec.Template.Spec.Domain.Devices.Interfaces[1].MacAddress)
+				Expect(err).ToNot(HaveOccurred())
+
+				updatedVM := *vm.DeepCopy()
+				By("before the NIC is deleted")
+				newVm := CreateVmObject(TestNamespace, false, []kubevirtv1.Interface{newInterface("br1", mac1.String())},
+					[]kubevirtv1.Network{newNetwork("br1")})
+				Eventually(func() error {
+					return testClient.VirtClient.Create(context.TODO(), newVm)
```
Comment (SchSeba) [inline tests/virtual_machines_test.go]: please validate the error here. we need to check that the error is related to the mac allocation and not something else

File: tests/virtual_machines_test.go [hunk 6]
```diff
@@ -489,6 +489,63 @@ var _ = Describe("Virtual Machines", func() {
...
+				Eventually(func() error {
+					return testClient.VirtClient.Create(context.TODO(), vm)
+
+				}, 50*time.Second, 5*time.Second).Should(Not(HaveOccurred()), "failed to apply the new vm object")
+				mac1, err := net.ParseMAC(vm.Spec.Template.Spec.Domain.Devices.Interfaces[0].MacAddress)
+				Expect(err).ToNot(HaveOccurred())
+				mac2, err := net.ParseMAC(vm.Spec.Template.Spec.Domain.Devices.Interfaces[1].MacAddress)
+				Expect(err).ToNot(HaveOccurred())
+
+				updatedVM := *vm.DeepCopy()
```
Comment (SchSeba) [inline tests/virtual_machines_test.go]: you don't need to Copy here.

just create the object
```golang
var updatedVM VirtualMachine
```

File: tests/virtual_machines_test.go [hunk 7]
```diff
@@ -489,6 +489,63 @@ var _ = Describe("Virtual Machines", func() {
...
+					[]kubevirtv1.Network{newNetwork("br1")})
+				Eventually(func() error {
+					return testClient.VirtClient.Create(context.TODO(), newVm)
+
+				}, 50*time.Second, 5*time.Second).Should(HaveOccurred(), "should have failed to create the vm because the mac is allready taken")
+
+				By("deleting the VM's NIC")
+				Eventually(func() error {
+					err := testClient.VirtClient.Get(context.TODO(),
+						client.ObjectKey{Namespace: updatedVM.Namespace, Name: updatedVM.Name}, &updatedVM)
```
Comment (SchSeba) [inline tests/virtual_machines_test.go]: use here the vm object not the update

File: tests/virtual_machines_test.go [hunk 8]
```diff
@@ -489,6 +489,63 @@ var _ = Describe("Virtual Machines", func() {
...
+					if err != nil {
+						return err
+					}
+					updatedVM.Spec.Template.Spec.Domain.Devices.Interfaces = []kubevirtv1.Interface{newInterface("br2", mac2.String())}
```
Comment (SchSeba) [inline tests/virtual_machines_test.go]: add a new line here

File: tests/virtual_machines_test.go [hunk 9]
```diff
@@ -489,6 +489,63 @@ var _ = Describe("Virtual Machines", func() {
...
+					updatedVM.Spec.Template.Spec.Networks = []kubevirtv1.Network{newNetwork("br2")}
+					err = testClient.VirtClient.Update(context.TODO(), &updatedVM)
+					if err != nil {
+						return err
+					}
+
+					if len(updatedVM.Spec.Template.Spec.Domain.Devices.Interfaces) != 1 {
+						return fmt.Errorf("failed to delete the VM's interface")
+					}
+					return nil
```
Comment (SchSeba) [inline tests/virtual_machines_test.go]: add new line here

File: tests/virtual_machines_test.go [hunk 10]
```diff
@@ -489,6 +489,63 @@ var _ = Describe("Virtual Machines", func() {
...
+
```
Comment (SchSeba) [inline tests/virtual_machines_test.go]: remove this empty line

File: tests/virtual_machines_test.go (hunk 2 above)
Comment (SchSeba) [reply tests/virtual_machines_test.go]: oh you are right miss that we have vm, newVm and updateVM haha

File: tests/virtual_machines_test.go [hunk 11]
```diff
@@ -489,6 +489,63 @@ var _ = Describe("Virtual Machines", func() {
...
+				}, 50*time.Second, 5*time.Second).Should(Not(HaveOccurred()), "failed to update client")
+
+				By("after the VM's NIC has been deleted")
+				Expect(updatedVM.Spec.Template.Spec.Domain.Devices.Interfaces[0].MacAddress).To(Equal(mac2.String()),
+					"kubemacpool should assign the same mac as before the update")
+				Eventually(func() error {
+					return testClient.VirtClient.Create(context.TODO(), newVm)
+
+				}, 50*time.Second, 5*time.Second).ShouldNot(HaveOccurred(), "failed to create the new VM")
+				Expect(newVm.Spec.Template.Spec.Domain.Devices.Interfaces[0].MacAddress).To(Equal(mac1.String()), "kubemacppool should assign the freed mac")
```
Comment (SchSeba) [inline tests/virtual_machines_test.go]: freed -> released

File: tests/virtual_machines_test.go [hunk 12]
```diff
@@ -489,6 +489,67 @@ var _ = Describe("Virtual Machines", func() {
...
+					if len(vm.Spec.Template.Spec.Domain.Devices.Interfaces) != 1 {
+						return fmt.Errorf("failed to delete the VM's interface")
+					}
+
+					return nil
+				}, 50*time.Second, 5*time.Second).Should(Not(HaveOccurred()), "failed to update client")
+
+				By("after the VM's NIC has been deleted")
+				Expect(vm.Spec.Template.Spec.Domain.Devices.Interfaces[0].MacAddress).To(Equal(mac2.String()),
+					"kubemacpool should assign the same mac as before the update")
```
Comment (phoracek) [inline tests/virtual_machines_test.go]: This is not a good test. It assumes that we always use the first available mac from the pool, and that is undocumented and can change any time. I mean, it does the job, but is not really what we want to test. Could we make it, so our pool has only 2 available addresses, if one from removed NIC has been recycled, we should be able to create a VM here and get **a** MAC. Sorry if I'm nitpicking too much.

Comment (alonSadan) [reply tests/virtual_machines_test.go]: HI @phoracek .It's all good :)
I didn't understand though whats the point of limiting the range if we are not checking kubemacpool assigns the same mac that was released (the only one that is available).
If we are checking that the new Vm received just "a MAC" than maybe we can leave the range as it is? 
Thanks:)

Comment (phoracek) [reply tests/virtual_machines_test.go]: This test schould check "do we release MAC after this and that happens with a VM". What you wrote checks that, but also "we always assign the first MAC address from the pool". The latter check is not documented and irrelevant.

I was thinking how could we check 1. without 2. What came to my mind is, let's limit the pool to two MAC addresses during the deployment. Then if the MAC is not released, we fail to allocate, if we release, we will have a MAC address to assign. Same results, less assumptions.

Does that make sense? Trying to do BDD testing, hehe.

Anyway, right now we don't control the range while testing, do we? It is only configured once during the deployment, isn't it?

Comment (SchSeba) [reply tests/virtual_machines_test.go]: we can do it

```golang
err := setRange(rangeStart, rangeEnd)
```

 we can put a range just before any test

Comment (alonSadan) [reply tests/virtual_machines_test.go]: > Does that make sense? Trying to do BDD testing, hehe.
OK. I see your point here.
NP I'll change that.
> Anyway, right now we don't control the range while testing, do we? It is only configured once during the deployment, isn't it?
I think we can. The "setRange" function sets the range for the macpool by configuring the range and deleting the kubemacpool pod and waiting for it to run again.

Comment (phoracek) [reply tests/virtual_machines_test.go]: I think it would be the right way to do it.

**Comment by alonSadan** [comment]
Hi @phoracek .
what do you think?

File: tests/virtual_machines_test.go [hunk 13]
```diff
@@ -488,6 +488,64 @@ var _ = Describe("Virtual Machines", func() {
...
+				Eventually(func() error {
+					return testClient.VirtClient.Create(context.TODO(), vm)
+
+				}, 50*time.Second, 5*time.Second).ShouldNot(HaveOccurred(), "failed to apply the new vm object")
+				_, err = net.ParseMAC(vm.Spec.Template.Spec.Domain.Devices.Interfaces[0].MacAddress)
+				Expect(err).ToNot(HaveOccurred())
+				_, err = net.ParseMAC(vm.Spec.Template.Spec.Domain.Devices.Interfaces[1].MacAddress)
+				Expect(err).ToNot(HaveOccurred())
+
+				By("before the NIC is deleted")
```
Comment (phoracek) [inline tests/virtual_machines_test.go]: This message won't be too helpful in test output. It should describe the following block in format "By checking that X is/does Y when Z"

File: tests/virtual_machines_test.go [hunk 14]
```diff
@@ -488,6 +488,64 @@ var _ = Describe("Virtual Machines", func() {
...
+					if err != nil {
+						return err
+					}
+
+					if len(vm.Spec.Template.Spec.Domain.Devices.Interfaces) != 1 {
+						return fmt.Errorf("failed to delete the VM's interface")
+					}
+
+					return nil
+				}, 50*time.Second, 5*time.Second).Should(Not(HaveOccurred()), "failed to update client")
```
Comment (phoracek) [inline tests/virtual_machines_test.go]: .ShouldNot(HaveOccurred())

Comment (phoracek) [inline tests/virtual_machines_test.go]: Update client? It updates VM's spec, what do you mean by "client"?

File: tests/virtual_machines_test.go [hunk 15]
```diff
@@ -488,6 +488,64 @@ var _ = Describe("Virtual Machines", func() {
...
+
+				By("after the VM's NIC has been deleted")
+				Eventually(func() error {
+					return testClient.VirtClient.Create(context.TODO(), newVM)
+
```
Comment (phoracek) [inline tests/virtual_machines_test.go]: please drop the empty line

File: tests/virtual_machines_test.go [hunk 16]
```diff
@@ -488,6 +488,64 @@ var _ = Describe("Virtual Machines", func() {
...
+					}
+
+					if len(vm.Spec.Template.Spec.Domain.Devices.Interfaces) != 1 {
+						return fmt.Errorf("failed to delete the VM's interface")
+					}
+
+					return nil
+				}, 50*time.Second, 5*time.Second).Should(Not(HaveOccurred()), "failed to update client")
+
+				By("after the VM's NIC has been deleted")
```
Comment (phoracek) [inline tests/virtual_machines_test.go]: ditto, should describe what you are checking

File: tests/virtual_machines_test.go (hunk 14 above)
Comment (alonSadan) [reply tests/virtual_machines_test.go]: oops...

**Pull request merged**